optics-core/src/Optics/Plated.hs

```diff
@@ -0,0 +1,352 @@
+-- Transformation
+-------------------------------------------------------------------------------
+
+-- | Transform every element in the tree, in a bottom-up manner.
+--
+-- For example, replacing negative literals with literals:
+--
+-- @
+-- negLits = 'transform' $ \\x -> case x of
+--   Neg (Lit i) -> Lit ('negate' i)
+--   _           -> x
+-- @
+transform :: Plated a => (a -> a) -> a -> a
+transform = transformOf plate
+{-# INLINE transform #-}
+
+-- | Transform every element in the tree in a bottom-up manner over a region
+-- indicated by a 'Setter'.
+transformOn :: (Is k A_Setter, Plated a) => Optic k is s t a a -> (a -> a) -> s -> t
+transformOn b = over b . transform
+{-# INLINE transformOn #-}
+
+-- | Transform every element by recursively applying a given 'Setter' in a
+-- bottom-up manner.
+transformOf :: Is k A_Setter => Optic k is a b a b -> (b -> b) -> a -> b
+transformOf l f = go where
+  go = f . over l go
+{-# INLINE transformOf #-}
+
+-- | Transform every element in a region indicated by a 'Setter' by recursively
+-- applying another 'Setter' in a bottom-up manner.
+transformOnOf
+  :: (Is k1 A_Setter, Is k2 A_Setter)
+  => Optic k1 is1 s t a b
+  -> Optic k2 is2 a b a b
+  -> (b -> b) -> s -> t
+transformOnOf b l = over b . transformOf l
+{-# INLINE transformOnOf #-}
+
+-- | Transform every element in the tree, in a bottom-up manner, monadically.
+transformM :: (Plated a, Monad m) => (a -> m a) -> a -> m a
+transformM = transformMOf plate
+{-# INLINE transformM #-}
+
+-- | Transform every element in the tree in a region indicated by a supplied
+-- 'Traversal', in a bottom-up manner, monadically.
+transformMOn
+  :: (Is k A_Traversal, Plated a, Monad m)
+  => Optic k is s t a a
+  -> (a -> m a) -> s -> m t
+transformMOn b = traverseOf b . transformM
+{-# INLINE transformMOn #-}
+
+-- | Transform every element in a tree using a user supplied 'Traversal' in a
+-- bottom-up manner with a monadic effect.
+transformMOf
+  :: (Is k A_Traversal, Monad m)
+  => Optic k is a b a b
+  -> (b -> m b) -> a -> m b
+transformMOf l f = go
+  where
+    go t = traverseOf l go t >>= f
+{-# INLINE transformMOf #-}
+
+-- | Transform every element in a tree that lies in a region indicated by a
+-- supplied 'Traversal', walking with a user supplied 'Traversal' in a bottom-up
+-- manner with a monadic effect.
+transformMOnOf
+  :: (Is k1 A_Traversal, Is k2 A_Traversal, Monad m)
+  => Optic k1 is1 s t a b
+  -> Optic k2 is2 a b a b
+  -> (b -> m b) -> s -> m t
+transformMOnOf b l = traverseOf b . transformMOf l
+{-# INLINE transformMOnOf #-}
+
+-------------------------------------------------------------------------------
+-- Paramorphisms
+-------------------------------------------------------------------------------
+
+-- | Perform a fold-like computation on each value, technically a paramorphism.
+paraOf :: Is k A_Fold => Optic' k is a a -> (a -> [r] -> r) -> a -> r
```

Comment:
I don't know to what degree we want to mimic the exact structure of `lens`, but arguably it would make sense to put the `*Of` variants in the modules for the optics they use, e.g. `paraOf` in `Fold`, `traverseOf` in `Setter` etc, leaving this module to contain the versions that actually use `Plated`.

(Which I just noticed is exactly how it is for `partsOf` already.)

Reply:
Indeed. I moved `paraOf` to `Optics.Fold` and removed `para` and `parts` for now.

Reply:
Oh, I didn't read this suggestion well enough it seems. Yeah, having `Optics.Plated` only have the functions with `plate` applied also sounds reasonable.